[PP] pt-native input/weight grad split

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134014
* #134052
* #133467
* __->__ #132691

Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[draft] [pp] pytorch native B-W grad split

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_schedule_multiproc.py:
```diff
@@ -383,6 +383,70 @@ def test_grad_with_manual_interleaved(self, ScheduleClass):
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize("ScheduleClass", [ScheduleWithW])
+    def test_schedule_with_native_zero_bubble(self, ScheduleClass):
```
looks like this is missing output/grad check, any reason not to just use the existing grad test?

H-Huang replied:
Finally fixed the grad bug and added a grad check.

I will clean up schedule args and add a new schedule specifically for zero-bubble to consolidate the tests in a followup PR

wconstab replied:
my question is whether there is a reason to duplicate the code of the test.  Is it identical to the other test_schedule that is already parametrized?  (should we just delete this test, and add `[ScheduleWithW, ScheduleFlexibleInterleaved1F1B]` to the other test, or is it useful to separate them?)

H-Huang replied:
Yeah the test has a lot of duplicate code, but also the schedule args are different so i would have to add if-statements in the `test_grad_with_manual_interleaved`. I think the test can be deleted but and added as one the parameterized tests, but I will do it as cleanup after I add the official zero bubble schedule

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -606,26 +613,42 @@ def backward_one_chunk(
+
+                # print(f"{bwd_kwargs['stage_output']}, {bwd_kwargs['input_values']}")
+                dinputs, param_groups = stage_backward_input(bwd_kwargs["stage_output"], bwd_kwargs["input_values"], self.submod.parameters())
+                self.backward_state[bwd_chunk_id] = (dinputs, input_values, param_groups)
```
nit: maybe just create a dw_runner here (lambda that calls stage_backward_weight with the right param groups..) and push it onto the self.dw_runner queue. then in W execution you dont have to check which mode you're in.

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -575,8 +582,8 @@ def backward_one_chunk(
-        if not full_backward:
-            assert self.dw_builder, "Must provide dw_builder to run partial backward"
+        # if not full_backward:
+        #     assert self.dw_builder, "Must provide dw_builder to run partial backward"
```
yup, delete this, and i think there might be a comment that needs to be updated explaining this is now optional override

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] [pp] pytorch native B-W grad split"


`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] pt-native input/weight grad split"


Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] pt-native input/weight grad split"


Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] pt-native input/weight grad split"


Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_backward.py:
```diff
@@ -53,6 +57,127 @@ def test_stage_backward(self):
+        ref_loss = loss_fn(ref_out, ref_target)
+        ref_loss.backward()
+
+        torch.testing.assert_close(x.grad, ref_x.grad)
```
nit: would be worth also asserting that mod.parameters grads are None?

H-Huang replied:
good idea, added

wconstab commented on test/distributed/pipelining/test_backward.py:
```diff
@@ -53,6 +57,127 @@ def test_stage_backward(self):
+        )
+
+        # backward of loss with respect to weights
+        dweights = stage_backward_weight(mod.parameters(), param_groups)
```
jw, is there a requirement that users call backward_input before backward_weight?

or if they just call backward_weight, will that work too?  (and if they subsequently call backward_input, would that reuse the intermediates from backward_weight?)

This doesn't really matter (i dont expect people to want to call in the wrong order) but might be a factor for 'api hardening'

H-Huang replied:
To call backward_weight the user would need the param groups which is outputted from backward_input. Theoretically they could get the parameter groups separately and just call backward_weight on its own.

> if they just call backward_weight, will that work too? (and if they subsequently call backward_input, would that reuse the intermediates from backward_weight?)

Currently no, we dont have support in backward_input to use the intermediates.

----- 
For the framework itself, I updated the error messaging so that if they are using pipeline stages (e.g. building their own schedules), then they will get an assert error: 

"Attempted to run backward_weight_one_chunk for chunk {bwd_chunk_id} without first calling `backward_one_chunk(full_backward=False)`"

and same for if they are using the dw_builder()

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+        self.node = node
+
+
+# TODO: use weak references to avoid reference cycle
```
what about this?

soulitzer replied:
This should be outdated, but probably worth double checking for no leaks in a small test

H-Huang replied:
Will follow up with a test

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+
+
+def _get_grad_fn_or_grad_acc(t: torch.Tensor) -> Union[Node, None]:
+    if t.requires_grad and t.grad_fn is None:
```
can we have comments about the overall behavior of this fn? why does takign a view get us a grad_fn when there is none? what does none mean? which users would fall into the RuntimeError path?

cc @soulitzer

soulitzer replied:
In the requires_grad=True and grad_fn is None path, since accumulate grad node is lazily created, we need to create a dummy view in order to trigger its creation. grad_fn of the view can be None if we are in a no-grad context.

This function as-is can also return None if t.requries_grad is False.

H-Huang replied:
Added comments, thanks Jeffrey!

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+            q.append(node)
+    while q:
+        node = q.popleft()
+        metadata = cast(Dict[str, List], node.metadata)
```
where does node.metadata actually come from? autograd gives it to us and we're just formally building a python weakref graph out of it? or we added it somewhere else?  would be good to comment the top of the function wth the overall semantics and why we want the reverse_closure

wconstab replied:
ok we add it below. and i guess 'reverse closure' refers to a closure over a reversed graph?

soulitzer replied:
> where does node.metadata actually come from? autograd gives it to us and we're just formally building a python weakref graph out of it? or we added it somewhere else?

autograd gives it to us

H-Huang replied:
Will add comments. autograd give it to us and we use it to traverse the graph. I need to do the casting because `Node` in graph.py has it defined as a method (https://github.com/pytorch/pytorch/blob/main/torch/autograd/graph.py#L77-L80), whereas it's used as just a dict, maybe that should be updated?

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+        metadata = cast(Dict[str, List], node.metadata)
+        reverse_edges = metadata.get("reverse_edges", [])
+        for holder_ref, idx in reverse_edges:
+            ref = holder_ref()
```
wouldn't hurt to add a mypy typedef for ref.  since i am finding myself wondering if its a Holder instance (but it cant be bc Holder doesn't support __call__()

H-Huang replied:
Have a work in progress PR that removes this holder_ref so I will hold off on this

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+        node = q.popleft()
+        for fn, idx in node.next_functions:
+            if fn is not None:
+                # Don't necessarily need to store on the graph
```
any downsides to storing it on the graph @soulitzer

wconstab replied:
oh, i think i'm starting to see whats going on.
1- we are storing new reverse pointer metadata on the original graph nodes
2- the reverse pointers are 2 stage for a reason: (a) we store weakrefs (to nodes, or to holders probably doesn't matter?) on the autograd nodes, so we dont create refcycles there.  (b) we also create strong refs to the nodes and hold onto those at least until we're done messing around with this graph, later we delete these to clean things up

Is that all correct?  this should be explained in comments

wconstab replied:
it does feel kinda bad to store stuff on the autograd graph.  after our code runs, the graph is mutated.  someone could notice (?) 

if we didn't want to modify the graph, i guess the alternative is to construct a ReverseGraph object, store a dict of id(node) -> node in it, and then manage the lifetime of this ReverseGraph object, making sure to delete it once we are done with it.  would that work?

soulitzer replied:
> it does feel kinda bad to store stuff on the autograd graph. after our code runs, the graph is mutated. someone could notice (

agreed, would be better to avoid using the grad_fn.metadata, and a separate ReverseGraph should work yeah

H-Huang replied:
In a follow up PR i am working on I just pass a dictionary to save this state as you mentioned!

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+                reverse_edges = metadata.get("reverse_edges", [])
+                if len(reverse_edges) == 0:
+                    q.append(fn)
+                holder = Holder(node)
```
probably a naive question, but why can't you directly do weakref.ref(node)?

soulitzer replied:
grad_fn doesn't support weak refs today and it may be tricky to add cc @albanD

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+    return reverse_graph_refs
+
+
+def get_param_groups(inputs: List[Node], params: List[Node]) -> List[Dict[str, Any]]:
```
should add a docstring for this

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+
+    # The assert will only be true if the input tensor requires gradients,
+    # otherwise the autograd graph will miss the first layer of inputs
+    # assert union_params == set(params)
```
was this assert causing a problem?

soulitzer replied:
We should probably keep the assert and just not call this for the first layer

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+
+
+def get_param_groups(inputs: List[Node], params: List[Node]) -> List[Dict[str, Any]]:
+    inputs_closure, _ = reverse_closure(inputs, set())
```
for my intuition.. this would be the reverse graph that starts with inputs, and goes up to the dOutput or the loss, but omits weights and any subgraphs connecting weights to this closure.  is that right?

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+                continue
+            closure.add(fn)
+            q.append(fn)
+    return closure, actual_target_nodes
```
i dont have a good intuition for what 'actual_target_nodes' a.k.a intersect represents..

soulitzer replied:
"target node" means if we see a node in this list, we stop traversal early

"actual target node" means during traversal these are the nodes in the set of target nodes we actually saw

maybe better renamed to "seen/visited_target_node"?

wconstab commented on torch/distributed/pipelining/_backward.py:
```diff
@@ -1,12 +1,234 @@
+            # save their inputs.
+            intermediate.register_prehook(get_hook(param_group, i))
+
+    # Stage 0 inputs do not require grads? Should we skip in that case?
```
well, if stage0's B operation is a no-op and its W operation is the whole W from the beginning, its not ideal.  It might be better to compute some partial products (the intersection between the Weights closure and Inputs closure) here.  Can be a follow up optimization.

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -611,26 +610,66 @@ def backward_one_chunk(
+        if self.dw_builder:
+            self.grads_input = self.backward_maybe_with_nosync(bwd_kwargs)
+            if full_backward:
+                self.dw_builder()()
```
a note for later (maybe add a TODO)

its not always perf equivalent to run dInput() followed by dWeight() vs running full backward.  For instance, separating the two may require longer lifetimes for any intermediate tensors, and might prevent comm/computation overlap if there is TP comm inside dInput that could have overlapped with some of dWeight.  If dw_builder is supplied, we may want to change our semantics so we are allowed to ignore the 'dw_builder' and call full_backward directly when it is a full_backward op.

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -611,26 +610,66 @@ def backward_one_chunk(
+                    self.submod.parameters(), param_groups
+                )
+            else:
+                # TODO: figure out a better way to do this:
```
i made a comment inside stage_backward_input, we might be able to fix this todo as well if we make the input fn do the intersect compute and then finish it off during W. seems ok to leave for follow up PR

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -102,7 +102,7 @@ def __init__(
                 Default: `None`.
-            dw_builder (Optional[Callable[[], Callable[[], None]]): If provided, dw_runner is a builder function
+            dw_builder (Optional[Callable[[], Callable[..., None]]): If provided, dw_runner is a builder function
                 that will build a new dw_runner function that will run parts of module backward that were intentionally
```
... otherwise, a dw_runner will be generated automatically by traversing the autograd graph.  (See Note: Automatic dW/dInput partitioning)

You really should add such a note that explains the algorithm e2e

wconstab reviewed (commented):
Overall this looks good to me. I asked for a lot of additional docs and comments.  

One significant code change would be to not mutate the autograd graph.  I'm actually OK with landing first and then refactoring this.

H-Huang committed:
Update on "[PP] pt-native input/weight grad split"


Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang committed:
Update on "[PP] pt-native input/weight grad split"


Add `stage_backward_input` and `stage_backward_weight` functions to perform the weight updates for inputs and weights independently.

We still support `self.dw_builder` argument for a custom backward, but it has become optional. It takes a separate code path and cannot be used in conjuction with the native zero backward.

Added tests:
`python test/distributed/pipelining/test_schedule_multiproc.py -k test_schedule_with_native_zero_bubble`
`python test/distributed/pipelining/test_backward.py`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge
